### PR TITLE
Use std::exception_ptr to transmit and store exceptions

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR updates the Ice C++ source code to use exception_ptr to transmit and store exceptions.

It's consistent with our async callback API, where errors are reported via a `std::function<void(exception_ptr)>`.

### CODE REVIEW

File: cpp/src/Ice/LocatorInfo.cpp [hunk 1]
```diff
@@ -86,19 +79,12 @@ class AdapterRequest : public LocatorInfo::Request
...
                 {
-                    try
-                    {
-                        rethrow_exception(e);
-                    }
-                    catch(const Exception& ex)
-                    {
-                        request->exception(ex);
-                    }
+                    request->exception(e);
```
Comment (pepone) [inline cpp/src/Ice/LocatorInfo.cpp:82]: call std::move?

**Review by pepone — APPROVED** [review]
Looks good, I think there is a few places where we can call `std::move` to avoid copying but we can review that in a follow up PR.

File: cpp/src/IceLocatorDiscovery/PluginI.cpp [hunk 2]
```diff
@@ -372,56 +365,36 @@ Request::response(bool ok, const pair<const Ice::Byte*, const Ice::Byte*>& outPa
...
-            throw Ice::ObjectNotExistException(__FILE__, __LINE__);
-        }
-        catch(...)
-        {
-            _amdCB.second(current_exception());
-        }
+        _amdCB.second(make_exception_ptr(Ice::ObjectNotExistException(__FILE__, __LINE__)));
     }
-    catch(const Ice::Exception&)
+    catch (...)
```
Comment (externl) [inline cpp/src/IceLocatorDiscovery/PluginI.cpp:395]: I thought we preferred catching std::exception as the most base exception? Everything should be deriving from it.

**Review by externl — APPROVED** [review]

File: cpp/src/Glacier2Lib/SessionHelper.cpp [hunk 3]
```diff
@@ -476,25 +476,24 @@ class ConnectFailed : public Ice::DispatcherCall
 public:
 
     ConnectFailed(const Glacier2::SessionCallbackPtr& callback, const Glacier2::SessionHelperPtr& session,
-                  const Ice::Exception& ex) :
+                  std::exception_ptr ex) :
         _callback(callback),
         _session(session)
     {
-        _ex = ex.ice_clone();
+        _ex = ex;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Glacier2Lib/SessionHelper.cpp:483]: It's not clear to me what we should move exception_ptr. Are they even movable?

According to https://en.cppreference.com/w/cpp/error/exception_ptr, exception_ptr meets the requirements of NullablePointer, which does not include movable.

File: cpp/src/Ice/CommunicatorI.cpp [hunk 4]
```diff
@@ -61,9 +61,9 @@ CommunicatorFlushBatchAsync::flushConnection(const ConnectionIPtr& con, Ice::Com
         }
 
         virtual bool
-        exception(const Exception& ex)
+        exception(std::exception_ptr ex)
         {
-            _childObserver.failed(ex.ice_id());
+            _childObserver.failed(getExceptionId(ex));
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/CommunicatorI.cpp:66]: It's not clear to me what you want to move.

File: cpp/src/Ice/ConnectionI.cpp [hunk 5]
```diff
@@ -1071,11 +1082,15 @@ Ice::ConnectionI::asyncRequestCanceled(const OutgoingAsyncBasePtr& outAsync, con
         {
             if(_asyncRequestsHint->second == outAsync)
             {
-                if(dynamic_cast<const Ice::ConnectionTimeoutException*>(&ex))
+               try
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/ConnectionI.cpp:1085]: Fixed

File: cpp/src/Ice/WSTransceiver.cpp [hunk 6]
```diff
@@ -387,23 +387,34 @@ IceInternal::WSTransceiver::closing(bool initiator, const Ice::LocalException& r
...
     {
         _closingReason = CLOSURE_TOO_BIG;
     }
+    catch (const Ice::ProtocolException&)
+    {
+        _closingReason  = CLOSURE_PROTOCOL_ERROR;
+    }
+    catch (...)
+    {
+       // no closing reason
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/WSTransceiver.cpp:416]: I just translated the old code. With the old code, we don't set _closingReason unless the exception is one of the types we check.

File: cpp/src/IceLocatorDiscovery/PluginI.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/IceLocatorDiscovery/PluginI.cpp:395]: That's correct, we assume all exceptions derive from std::exception.

However, in this particular code, we don't care: the fallback/catchall is to forward the exception. I see this as similar to:
```
catch { ... }
```
vs
```
catch (Exception) { ... }
```
in C#. 

catch { } in C# corresponds to catch (...) in C++.

File: cpp/src/Ice/CommunicatorI.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (externl) [reply cpp/src/Ice/CommunicatorI.cpp:66]: The exception, but I think it's probably not worth trying to optimize these.

**Pull request merged**